pkg/rollup/rollup.go

```diff
@@ -0,0 +1,133 @@
+		}
+		lName := items[0]
+
+		var res util.DayDuration
+		err := res.UnmarshalText([]byte(items[1]))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing resolution %s: %w", items[1], err)
+		}
+
+		var retention util.DayDuration
+		err = retention.UnmarshalText([]byte(items[2]))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing retention %s: %w", items[2], err)
+		}
+		r = append(r, DownsampleResolution{lName, res, retention})
+	}
+	return r, nil
+}
+
+// EnsureRollupWith ensures "strictly" that the given new resolutions are applied in the database.
+// Note: It follows a "strict" behaviour, meaning any existing resolutions of downsampling in
+// the database will be removed, so that the all downsampling data in the database strictly
+// matches the provided newResolutions.
+func EnsureRollupWith(conn *pgx.Conn, newResolutions []DownsampleResolution) error {
+	rows, err := conn.Query(context.Background(), "SELECT name, resolution, retention FROM _prom_catalog.rollup")
+	if err != nil {
+		return fmt.Errorf("querying existing resolutions: %w", err)
+	}
+	defer rows.Close()
+
+	var existingResolutions []DownsampleResolution
+	for rows.Next() {
+		var lName string
+		var resolution, retention time.Duration
+		if err := rows.Scan(&lName, &resolution, &retention); err != nil {
+			return fmt.Errorf("error scanning output rows for existing resolutions: %w", err)
+		}
+		existingResolutions = append(existingResolutions, DownsampleResolution{label: lName, resolution: util.DayDuration(resolution), retention: util.DayDuration(retention)})
+	}
+
+	// Determine which resolutions need to be created and deleted from the DB.
+	pendingCreation := diff(newResolutions, existingResolutions)
+	pendingDeletion := diff(existingResolutions, newResolutions)
+
+	// Delete rollups that are no longer required.
+	if err = deleteRollups(conn, pendingDeletion); err != nil {
+		return fmt.Errorf("delete rollups: %w", err)
+	}
+
+	// Create new rollups.
+	if err = createRollups(conn, pendingCreation); err != nil {
+		return fmt.Errorf("create rollups: %w", err)
+	}
+	return nil
+}
+
+func createRollups(conn *pgx.Conn, res []DownsampleResolution) error {
+	for _, r := range res {
+		_, err := conn.Exec(context.Background(), "CALL _prom_catalog.create_metric_rollup($1, $2, $3)", r.label, time.Duration(r.resolution), time.Duration(r.retention))
+		if err != nil {
+			return fmt.Errorf("error creating rollup for %s: %w", r.label, err)
+		}
+	}
+	return nil
+}
+
+func deleteRollups(conn *pgx.Conn, res []DownsampleResolution) error {
+	for _, r := range res {
+		_, err := conn.Exec(context.Background(), "CALL _prom_catalog.delete_metric_rollup($1)", r.label)
+		if err != nil {
+			return fmt.Errorf("error deleting rollup for %s: %w", r.label, err)
+		}
+	}
+	return nil
+}
+
+// diff returns the elements of a that are not in b.
```

Comment:
Ok, so here are look for a complement of set B. Can you explain more in comments why this is needed.

Reply:
I think it would be nice mentioning here why we need this.